code/modules/power/turbine/turbine.dm

```diff
@@ -0,0 +1,557 @@
+		damage = max(damage - TURBINE_DAMAGE_HEALING, 0)
+
+	if((damage - damage_archived >= 2 || damage > TURBINE_DAMAGE_ALARM_START) && COOLDOWN_FINISHED(src, turbine_damage_alert))
+		damage_alert(damage_done)
+
+/obj/machinery/power/turbine/core_rotor/proc/damage_alert(damage_done)
+	COOLDOWN_START(src, turbine_damage_alert, max(round(TURBINE_DAMAGE_ALARM_START - damage_done), 5) SECONDS)
+
+	var/integrity = damage / 500
+	integrity = max(round(100 - integrity * 100, 0.01), 0)
+
+	if(integrity <= 0)
+		failure()
+		return
+
+	radio.talk_into(src, "Warning, turbine at [get_area_name(src)] taking damage, current integrity at [integrity]%!", engineering_channel)
+
+/obj/machinery/power/turbine/core_rotor/proc/failure()
+	deactivate_parts()
+	if(rpm < 35000)
+		explosion(src, 0, 1, 4)
+		return
+	if(rpm < 87500)
+		explosion(src, 0, 2, 6)
+		return
+	if(rpm < 220000)
+		explosion(src, 1, 3, 7)
+		return
+	if(rpm < 550000)
+		explosion(src, 2, 5, 7)
+
+/obj/machinery/power/turbine/core_rotor/process_atmos()
+
+	if(!active || !all_parts_connected)
+		return
+
+	var/datum/gas_mixture/input_turf_mixture = input_turf.air
+
+	if(!input_turf_mixture || !input_turf_mixture.gases)
+		return
+
+	calculate_damage_done(input_turf_mixture.temperature)
+
+	var/compressor_work = input_turf_mixture.total_moles() * R_IDEAL_GAS_EQUATION * input_turf_mixture.temperature * log(input_turf_mixture.volume / compressor_mixture.volume) * TURBINE_WORK_CONVERSION_MULTIPLIER
+	input_turf.air.pump_gas_to(compressor_mixture, input_turf.air.return_pressure())
+	input_turf.air_update_turf(TRUE)
+	compressor_mixture.temperature = max((compressor_mixture.temperature * compressor_mixture.heat_capacity() + compressor_work * compressor_mixture.total_moles() * TURBINE_HEAT_CONVERSION_MULTIPLIER) / compressor_mixture.heat_capacity(), TCMB)
+
+	var/compressor_pressure = compressor_mixture.return_pressure()
+
+	var/rotor_work = compressor_mixture.total_moles() * R_IDEAL_GAS_EQUATION * compressor_mixture.temperature * log(compressor_mixture.volume / rotor_mixture.volume) * TURBINE_WORK_CONVERSION_MULTIPLIER
+	rotor_work = rotor_work - compressor_work
+	compressor_mixture.pump_gas_to(rotor_mixture, compressor_mixture.return_pressure())
+	rotor_mixture.temperature = max((rotor_mixture.temperature * rotor_mixture.heat_capacity() + rotor_work * rotor_mixture.total_moles() * TURBINE_HEAT_CONVERSION_MULTIPLIER) / rotor_mixture.heat_capacity(), TCMB)
+
+	var/turbine_work = rotor_mixture.total_moles() * R_IDEAL_GAS_EQUATION * rotor_mixture.temperature * log(rotor_mixture.volume / turbine_mixture.volume) * TURBINE_WORK_CONVERSION_MULTIPLIER
+	turbine_work = turbine_work - abs(rotor_work)
+	rotor_mixture.pump_gas_to(turbine_mixture, rotor_mixture.return_pressure())
+	turbine_mixture.temperature = max((turbine_mixture.temperature * turbine_mixture.heat_capacity() + turbine_work * turbine_mixture.total_moles() * TURBINE_HEAT_CONVERSION_MULTIPLIER) / turbine_mixture.heat_capacity(), TCMB)
+
```

Comment:
Sort of on the topic of moving these gasmixture datums onto their machine owners, I get the feeling using procs for most of this would help clean up the logic a bit. at least getting the work values and such

Reply:
Aight will work to move stuff on the right place